### ISSUE TITLE
Rust Client Idiomatic Code

### ISSUE THREAD

**Issue by adam-bates** [Vexilla/vexilla#5]
Hey Chris, It's Snowy (Just like the last issue, I'm not worried about a dox!)

I have a collection of minor suggestions for more "idiomatic" rust code. None of them seem like they deserve their own issue, so I'm going to clump here. If you'd prefer me to split to separate issues, I'm happy to 😄 

1. `impl Into<&'static str>`: static strs are essentially string literals (ignoring intentional memory leaking). Which means it doesn't really make sense to take a type that converts to a string literal. If you definitely need the str to be static, then I would just use `&'static str` without the impl Into.

2. `VexillaClient` is storing 3 `&'static str` fields. This forces developers to use string literals, rather than for example a format string, or something read in from an environment file. I would suggest using the owned `String`. If you want to keep the nice API, your `VexillaClient::new` function could accept `impl Into<String>`
    
    Note: If you're trying to avoid unnecessary clones of `String`, you could use a `Cow<'static, str>` (Clone-On-Write), or `Rc<str>` (Reference-counted).

3. `unwrap` is unpopular in Rust libraries because developers like to be in charge of when they handle errors and when they panic. There are a couple of spots in the library code that unwrap when they could just return a result with the error.

4. `group_lookup_table[scrubbed_file_name.as_str()]` can panic if there are no matches in the table. The "rusty" way would be to call `.get(...)` on `group_lookup_table`, and dealing with the situation where there is no match.

5. Enums and Clones. So in Rust, if I have the following enum:
    ```rust
    enum Example {
        LotsOfMem(TonsOfData), // 1kb of stack data
        NotMuchMem(u8),
    }
    ```
    
    and I create `Example::NotMuchMem(0)`, this will actually allocate 1kb of memory on the stack. And cloning it will allocate another 1kb of memory on the stack. Even though I definitely don't need that. Instead, we can `Box` the data to put it on the heap, keeping our stack allocation small. Or better yet, `Rc` the data to put it on the heap AND prevent cloning huge amounts of data unnecessarily.

    There's some code in your Rust client that could probably benefit from this change. I'm mostly focused on the code that checks feature flags at runtime as this is usually where developers care about cpu and memory costs. 

These are all pretty minor. The client looks awesome. Hope you enjoyed the rust coding 😄 

I'm more than happy to make a commit addressing any / all of these if you're interested.

**Comment by cmgriffing** [comment]
Thanks. Sorry for the delay. I think I have addressed a bunch of it. Would it be ok to create smaller separate issues for anything I might have missed?

1) We need to use this because we have the ability to generate enums for our users using a "type generation" tool that is located in the `cli` package. To work with those enums, we needed the Into (as far as I know).

2) Done. I implemented the Into. I will publish the changes soon.

3) I believe I have removed any `unwrap` that is outside of a test. In a test, I think it is acceptable.

4) Looks like I only had it left in one place. I believe I have fixed it now.

5) I think my understanding is lacking enough that I would love a separate issue for this and then we can have some more specific examples and links.
